### DISCUSSION TITLE
Bookmark actions

### DISCUSSION THREAD

**Discussion by yok0** [pymupdf/PyMuPDF#1301] [Looking for help]
I just have a question as I went through the docs back and forward and could not find this option.

I have a big pdf with lots of bookmarks. Unfortunately when bookmark is clicked the pdf changes it's zoom to "Fit Page"
and it must be 100%. The way we do it now is to change initial view to 100% and set bookmark action to "Inherit zoom".
This works perfectly but changing hundreds of bookmarks manually is just ridiculous. 

Does PyMuPDF allows change of these properties ?

![image](https://user-images.githubusercontent.com/6692754/135443977-6ea52b1a-ab6a-4cce-87ff-798b9361ac72.png)

![image](https://user-images.githubusercontent.com/6692754/135444081-bd7e65c1-a875-48c5-b15c-6d981988dd6f.png)

**Comment by JorjMcKie** [comment]
> Does PyMuPDF allows change of these properties?

In principle, yes (ever heard of Radio Erivan?).
The API is explained [here](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_toc_item).
As per view zoom level, I only support the PDF format ``/XYZ x y z``, where (x, y) is the top-left coordinate on the page and z is the zoom factor: 1 = 100%. Zoom = 0 means "not specified".
Let's see if that is ok for you.
Please do not hesitate to ask if you run into problems.

> I am going to move this to **_Discussions_**.

**Reply by yok0** [reply]
Thank you for you reply.

Unfortunately solution suggested does not work. I've put quickly this together.

```
import fitz

doc = fitz.open("bb.pdf")
toc = doc.get_toc(False)

for i, item in enumerate(toc):
    if i > 6:
        break
    index = item[0]
    title = item[1]
    to = item[2]
    opts_dict = item[3]

    opts_dict['zoom'] = 1.0
    print(f"Processing item : {i}")
    doc.set_toc_item(index, dest_dict=opts_dict)

doc.save("newaaa.pdf", garbage=3, deflate=True)
```

It does change `zoom` value in the dict but it has no effect on the pdf and the bookmark action still shows "Fit Page"
`{'kind': 1, 'xref': 62145, 'page': 65, 'to': Point(0.0, 0.0), 'zoom': 1.0, 'bold': True, 'color': (0.0, 0.0, 0.0)}`

I have also checked the zoom value on the working pdf and it is set to `0.0` Did not find any differences when it comes to ToC.
Seems to me like this setting may be hidden somewhere else. I'll keep on digging

**Reply by bexshay** [reply]
My New Year Greetings!

I tried to implement this functionality with my pdf document and it has the same symptoms as above.

I am using Name: PyMuPDF Version: 1.19.3
(below is amended example from [here](https://github.com/pymupdf/PyMuPDF/discussions/782))
"""
doc = fitz.open(src)
toc = doc.get_toc(simple=False)
item = toc[4]  # we want to modify item 'n'
d = item[3]  # the destination details dictionary of that item
d["color"] = (1,0,0)  # we want it red
d["italic"] = True  # and italic
d["bold"] = True  # and bold
d["zoom"] = 1.0
print(d)
doc.save("NEWSWIG.pdf")
"""
{'kind': 1, 'xref': 810, 'page': 5, 'to': Point(72.0, 36.0), 'zoom': 1.0, 'italic': True, 'bold': True, 'color': (1.0, 0.0, 0.0)}

Unfortunately, color and zoom in real pdf have not changed. May be, I am doing something wrong?

**Reply by Augus1999** [reply]
Before `doc.save("NEWSWIG.pdf")`, I think you need update your toc into the document via `doc.set_toc(toc)`. Hope this can help you.

**Reply by bexshay** [reply]
Thank you for your comment!
![image](https://user-images.githubusercontent.com/43413269/147978243-560a9676-88af-4bc1-ad64-977e84e61c3e.png)
Colors worked!
Regarding the Zoom level, with 1 it turned to "Actual Size", that is 100%, I believe.

In my version of Acrobat Pro DC there are 5 level options to select :
![image](https://user-images.githubusercontent.com/43413269/147978800-3437ef35-b5bb-4255-b837-bf468a88b34a.png)

Is there a work around to set zoom to be "Inherit Zoom"?

Also, I think local page bookmarks need separate treatment for Zoom level (again in scope of 0.0 or 1.0).

**Reply by yok0** [reply]
Yes, you need to set zoom level for every bookmark. I have not found a solution yet, if you do please let me know.
We went with a paid (cheap) solution, a small app written in Java and its only purpose is to change bookmarks zoom level to 'inherit'.

**Reply by ThoKau** [reply]
Hi,
I'm s struggling on generating a new toc from a list with set_toc(toc). Each entry has 3 parameters for level, page and title which works.
For some reason PyMuPDF doesn't set the destination to the page borger but adds a crappy annoying border of 72/36 (dots?).
I need a destination of  0.0/0.0 so I have to use the optional 4th parameter which is not documented.
It looks as mentioned a few posts before:
{'kind': 1, 'xref': 62145, 'page': 65, 'to': Point(0.0, 0.0), 'zoom': 1.0, 'bold': True, 'color': (0.0, 0.0, 0.0)}
The problem is the "point"-statement with this message:     "NameError: name 'Point' is not defined"
What is the correct syntax? Any packages to install or import? 
Thanks!

Edit:  from fitz import Point   seems to work.

**Reply by JorjMcKie** [reply]
> Hi, I'm s struggling on generating a new toc from a list with set_toc(toc). Each entry has 3 parameters for level, page and title which works. For some reason PyMuPDF doesn't set the destination to the page borger but adds a crappy annoying border of 72/36 (dots?). I need a destination of 0.0/0.0 so I have to use the optional 4th parameter which is not documented. It looks as mentioned a few posts before: {'kind': 1, 'xref': 62145, 'page': 65, 'to': Point(0.0, 0.0), 'zoom': 1.0, 'bold': True, 'color': (0.0, 0.0, 0.0)} The problem is the "point"-statement with this message: "NameError: name 'Point' is not defined" What is the correct syntax? Any packages to install or import? Thanks!

The "crappy annoying" parameter `(72, 36)` means a point (not a border) 1 inch from left, 0.5 inches from top. The fourth parameter **_is_** documented. You also could have just used a float, say 0, then Point(72,0) would have been taken. IAW use `[lvl, title, page, 0]`.
BTW if you use the name "Point", you must at least tell Python which package has a definition for this, IAW use `fitz.Point`.
If all these defaults don't meet your requirement, prepare a minimal template dictionary like `[lvl, title, page, {"kind":1, "page":page-1, "to": fitz.Point(0,0)}]` with overrides where you don't like the defaults.
Note that "xref" is always ignored in this context - it only is being inserted when you **read** the bookmark item from the file and then useful for singleton updates.

**Reply by Augus1999** [reply]
> Hi, I'm s struggling on generating a new toc from a list with set_toc(toc). Each entry has 3 parameters for level, page and title which works. For some reason PyMuPDF doesn't set the destination to the page borger but adds a crappy annoying border of 72/36 (dots?). I need a destination of 0.0/0.0 so I have to use the optional 4th parameter which is not documented. It looks as mentioned a few posts before: {'kind': 1, 'xref': 62145, 'page': 65, 'to': Point(0.0, 0.0), 'zoom': 1.0, 'bold': True, 'color': (0.0, 0.0, 0.0)} The problem is the "point"-statement with this message: "NameError: name 'Point' is not defined" What is the correct syntax? Any packages to install or import? Thanks!
> 
> Edit: from fitz import Point seems to work.

You need import the class "Point" via `from fitz import Point` first.

**Reply by monchin** [reply]
> I have not found a solution yet, if you do please let me know.

Hi, I have dug into this issue for some time, and found that you can set "Inherit Zoom" by changing dest_dict in toc items.

You should ensure that "kind"  = 1, "view" not in the dest_dict, "zoom" = 0, "page" = toc_item[2] - 1, and maybe no negative number in "to", and then after you `set_toc`, you'll see the bookmark is "inherit zoom".

But seems we cannot set bookmark as "fit width" or other options. If one bookmark item is "fit width" or others other than "inherit zoom", if we do
```python
toc = doc.get_toc(False)
doc.set_toc(toc)
with open('test.pdf', 'wb') as f:
    f.write(doc.write())
```
Bookmarks would become invalid, and you would find the new toc would be different with the old toc by `doc.get_toc(False)`. That's pretty strange, seems like a bug.
